src/gentropy/common/per_chromosome.py

```diff
@@ -0,0 +1,98 @@
+        tmp_variant_annotation_path (str): The path to store the temporary output for the repartitioned annotation dataset.
+    """
+    va_df = (
+        session
+        .spark
+        .read
+        .parquet(variant_annotation_path)
+    )
+    va_df_direct = (
+        va_df.
+        select(
+            f.col("chromosome").alias("vaChromosome"),
+            f.col("variantId"),
+            f.concat_ws(
+                "_",
+                f.col("chromosome"),
+                f.col("position"),
+                f.col("referenceAllele"),
+                f.col("alternateAllele")
+            ).alias("ukb_ppp_id"),
+            f.lit("direct").alias("direction")
+        )
+    )
+    va_df_flip = (
+        va_df.
+        select(
+            f.col("chromosome").alias("vaChromosome"),
+            f.col("variantId"),
+            f.concat_ws(
+                "_",
+                f.col("chromosome"),
+                f.col("position"),
+                f.col("alternateAllele"),
+                f.col("referenceAllele")
+            ).alias("ukb_ppp_id"),
+            f.lit("flip").alias("direction")
+        )
+    )
+    (
+        va_df_direct.union(va_df_flip)
+        .coalesce(1)
+        .repartition("vaChromosome")
+        .write
+        .partitionBy("vaChromosome")
+        .mode("overwrite")
+        .parquet(tmp_variant_annotation_path)
+    )
+
+
+def process_summary_stats_per_chromosome(session: SparkSession, ingestion_class: type[UkbPppEurSummaryStats], raw_summary_stats_path: str, tmp_variant_annotation_path: str, summary_stats_output_path: str) -> None:
+    """Processes summary statistics for each chromosome, partitioning and writing results.
+
+    Args:
+        session (SparkSession): The Spark session to use for distributed data processing.
+        ingestion_class (type[UkbPppEurSummaryStats]): The class used to handle ingestion of source data. Must have a `from_source` method returning a DataFrame.
+        raw_summary_stats_path (str): The path to the raw summary statistics files.
+        tmp_variant_annotation_path (str): The path to temporary variant annotation data, used for chromosome joins.
+        summary_stats_output_path (str): The output path to write processed summary statistics as parquet files.
+    """
+    # Set mode to overwrite for processing the first chromosome.
+    write_mode = "overwrite"
+    # Chromosome 23 is X, this is handled downstream.
+    for chromosome in list(range(1, 24)):
+        logging_message = f"  Processing chromosome {chromosome}"
+        session.logger.info(logging_message)
+        (
+            ingestion_class.from_source(
+                spark=session.spark,
+                raw_summary_stats_path=raw_summary_stats_path,
+                tmp_variant_annotation_path=tmp_variant_annotation_path,
+                chromosome=str(chromosome),
+            )
+            .df
+            .coalesce(1)
+            .repartition("studyId", "chromosome")
```

Comment:
You do quiet a big amount of shuffling. Also are you sure that coalescing the chromosome partitionned dataset to a single partition is not redundant in this case?

Reply:
That's a fair point, however there's a tradeoff to be made here. If we don't do the shuffling, then for each chromosome of each study we get multiple very small parquet files, which will (1) slow down downstream processing and (2) make it more expensive as we'll be incurring lots of costs for Google Cloud Storage (API operations). From what I observed when running a similar thing for UKB PPP, shuffling doesn't introduce too much of a noticeable tradeoff; so I'd propose we keep it this way